src/scipp/optimize/__init__.py

```diff
@@ -42,17 +43,29 @@ def _covariance_with_units(pcov, units):
                 u = uj
             elif uj is not None:
                 u = ui * uj
-            pcov[i, j] = _as_scalar(elem, u)
+            pcov[p_names[i]][p_names[j]] = _as_scalar(elem, u)
     return pcov
 
 
+def _make_defaults(f, p0):
+    spec = getfullargspec(f)
+    if len(spec.args) != 1 or spec.varargs is not None:
+        raise ValueError("Fit function must take exactly one positional argument")
```

Comment:
Do we really have to be so strict here? We can still call the function with keyword args even if it has positional args.
I think most users will not want to worry about putting `*,` into the arg list. Especially when using a lambda.

Reply:
I think the main reason is simplicity. Without this the handling of the params is complicated significantly, and it is unclear what we should return? What would `p0` be? A tuple of a list and dict? Same for `popt`, returning a tuple of a list and a dict that is hard to handle? And `pcov` would be even more complicated.

Reply:
I would always return dicts.

I just don't know if the users will be very happy with this requirement. But if you say it is best to require keyword args, we can keep it as it is and change later in case it becomes a problem. It would not be a breaking change.

Reply:
> I would always return dicts.

... and extra the names of position arguments from the signature?

> But if you say it is best to require keyword args, we can keep it as it is and change later in case it becomes a problem. It would not be a breaking change.

Agree, we can change this later easily (but not the other way round).

Reply:
> ... and extra the names of position arguments from the signature?

Yes